src/Main.elm

```diff
@@ -1,10 +1,209 @@
 module Main exposing (main)
 
-import Data.Audience
-import Data.AudienceFolder
-import Html exposing (Html)
+import Browser
+import Data.Audience as A
+import Data.AudienceFolder as F
+import Dict
+import Html
+import Html.Attributes as Hat
+import Html.Events as Hev
+import Json.Decode as Jd
+import Parser as P
+import Set
 
 
-main : Html msg
+main : Program () Model Msg
 main =
-    Html.text "There will be app soon!"
+    Browser.sandbox
+        { init = init
+        , view = view
+        , update = update
+        }
+
+
+type Msg
+    = FolderClick Int
+    | GoUpClick
+
+
+type alias Model =
+    Result String GoodModel
+
+
+{-| The integer IDs are not the same as the ones provided in the data.
+It seems likely that the provided IDs are row IDS from the database,
+and that folders and audiences are kept in different tables, so
+the IDs may well conflict. A globally unique ID is needed here.
+
+This model design is inspired by the book Data-Oriented Design by
+Richard Fabian, see <https://www.dataorienteddesign.com/dodbook> .
+It's an idea from game dev that says that using tabular data
+structures gives better performance. I am experimenting with using
+it in high-level languages, like Elm and Go. I am finding so far
+that it gives a really elegant and minimal internal state.
+
+-}
+type alias GoodModel =
+    -- The keys are child IDs and the values are parent IDs. Root
+    -- items are not included. If a map lookup fails then the item
+    -- is root, so there are no internal errors caused by failing map
+    -- lookups. The reason for using a dictionary rather than a list
+    -- is to enforce that a child can have only one parent.
+    { parents : Dict.Dict Int Int
+
+    -- As well as providing the AudienceType for audiences, this is
+    -- used to detect whether something is a folder or an audience.
+    , audiences : Dict.Dict Int A.AudienceType
+
+    -- The IDs and names for all the folders and audiences. This should
+    -- only be mapped over, not used for lookup. Lookup is bad in this
+    -- case because failure will have to be handled, and it shouldn't
+    -- ever be possible to have an ID without a name. The reason for
+    -- using a dictionary instead of a list is to enforce that an
+    -- item can only have one name.
+    , all : Dict.Dict Int String
```

Comment:
Yes I like this idea. And I just realised that if I assume names are unique within a folder, I don't need audience IDs at all, because an audience is uniquely identified by the combination of its name and its parent ID. So now I only have folder IDs and can use the original ones. In this new Model, I always store folder names alongside the folder IDs, because they always have a name, and this prevents having to look up names, which could fail. I can also drop 135 lines of complicated parser code, so hopefully it won't seem so over-engineered.